src/drivers/snapdragon_pwm_out/snapdragon_pwm_out.cpp

```diff
@@ -0,0 +1,608 @@
+	close(_fd);
+
+}
+
+void send_outputs_pwm(const uint16_t *pwm)
+{
+
+
+	/*
+	 * send pwm in us: ToDo: check if it is in us
+	 */
+	for (unsigned i = 0; i < NUM_PWM; ++i) {
+		_pwm[i].pulse_width_in_usecs = pwm[i];
+	}
+}
+
+
+
+void task_main(int argc, char *argv[])
+{
+	_is_running = true;
+
+	if (pwm_initialize(_device) < 0) {
+		PX4_ERR("Failed to initialize PWM.");
+		return;
+	}
+
+	// Set up mixer
+	if (initialize_mixer(MIXER_FILENAME) < 0) {
+		PX4_ERR("Mixer initialization failed.");
+		return;
+	}
+
+	_mixer->groups_required(_groups_required);
+	// subscribe and set up polling
+	subscribe();
+
+	// Start disarmed
+	_armed.armed = false;
+	_armed.prearmed = false;
+
+	// set max min pwm
+	pwm_limit_init(&_pwm_limit);
+
+	// Main loop
+	while (!_task_should_exit) {
+
+		/* wait up to 10ms for data */
+		int pret = px4_poll(_poll_fds, _poll_fds_num, 10);
+
+
+		/* Timed out, do a periodic check for _task_should_exit. */
+		if (pret == 0) {
+			continue;
+		}
+
+		/* This is undesirable but not much we can do. */
+		if (pret < 0) {
+			PX4_WARN("poll error %d, %d", pret, errno);
+			/* sleep a bit before next try */
+			usleep(10000);
+			continue;
+		}
+
+
+		/* get controls for required topics */
+		unsigned poll_id = 0;
+
+
+		for (uint8_t i = 0; i < actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS; i++) {
+
+			if (_controls_subs[i] >= 0) {
+
+				if (_poll_fds[poll_id].revents & POLLIN) {
+
+					orb_copy(_controls_topics[i], _controls_subs[i], &_controls[i]);
+
+				}
+
+				poll_id++;
+			}
+		}
+
+		if (_mixer != nullptr) {
+
+
+			_outputs.timestamp = hrt_absolute_time();
+
+			/* do  mixing for virtual control group */
+			_outputs.noutputs = _mixer->mix(_outputs.output,
+							_outputs.NUM_ACTUATOR_OUTPUTS,
+							NULL);
+
+
+			/* disable unused ports by setting their output to NaN */
+			for (size_t i = _outputs.noutputs; i < _outputs.NUM_ACTUATOR_OUTPUTS; i++){
+
+				_outputs.output[i] = NAN;
+
+			}
+
+			const uint16_t reverse_mask = 0;
+			uint16_t disarmed_pwm[4];
+			uint16_t min_pwm[4];
+			uint16_t max_pwm[4];
+
+			for (unsigned int i = 0; i < 4; i++) {
+				disarmed_pwm[i] = _pwm_disarmed;
+				min_pwm[i] = _pwm_min;
+				max_pwm[i] = _pwm_max;
+			}
+
+			uint16_t pwm[4];
+
+
+			// TODO FIXME: pre-armed seems broken -> copied and pasted from pwm_out_rc_in: needs to be tested
+			pwm_limit_calc(_armed.armed,
+				       false/*_armed.prearmed*/,
+				       _outputs.noutputs,
+				       reverse_mask,
+				       disarmed_pwm,
+				       min_pwm,
+				       max_pwm,
+				       _outputs.output,
+				       pwm,
+				       &_pwm_limit);
+
+
+			if (_armed.lockdown) {
+				send_outputs_pwm(disarmed_pwm);
+
+			} else {
+				send_outputs_pwm(pwm);
+			}
+
+			if (_outputs_pub != nullptr) {
+				orb_publish(ORB_ID(actuator_outputs), _outputs_pub, &_outputs);
+
+			} else {
+				_outputs_pub = orb_advertise(ORB_ID(actuator_outputs), &_outputs);
+			}
+
+
+		} else {
+			PX4_ERR("Could not mix output! Exiting...");
```

Comment:
Instead of having this long `if` `else`, I'd check:

```
if (_mixer == nullptr) {
    task_should_exit = true;
    continue;
}

...
```

This requires less indentation and is easier to read.